Print style view

sisiz opened the pull request:
Monitoring page with table that displays the print styles form the `print_style_recipes` table. Each print style can be clicked on and takes you to a page of all the books with that print_style. Each of those books then also links back to the content-status page for the baking history of that book (the page that is part of the other pr for the content status view)

The tests are commented out because of the celery globalization issue, and also because the test data doesn't have these tables filled in.

pumazi commented on cnxpublishing/tests/views/test_admin.py:
```diff
@@ -100,6 +101,81 @@ def test(self):
+
+        from ...views.admin import admin_print_styles_single
+        with self.assertRaises(httpexceptions.HTTPBadRequest) as cm:
+            admin_print_styles_single(request)
```
Wouldn't this likely be a 404 Not Found rather than 400 Bad Request?

pumazi commented on cnxpublishing/tests/views/test_admin.py:
```diff
@@ -100,6 +101,81 @@ def test(self):
+        for row in content['collections']:
+            self.assertEqual(set(row.keys()),
+                             set(['title', 'authors', 'revised', 'uuid',
+                                  'ident_hash', 'status']))
```
How about we test for values as well? Maybe use one of the use-cases and adjust the data in the database to include a print-style.

pumazi commented:
This branch has been force pushed. Remember to hard reset your local branch to make edits.

sisiz commented:
I fixed all the things requested above except the dict(zip(...)) because I added the route_path into the dict so the zipping didn't look as clean to me, but not sure what you think.

karenc commented on cnxpublishing/views/admin.py:
```diff
@@ -271,3 +274,86 @@ def admin_edit_site_message_POST(request):
+                AND portal_type='Collection'
+                ORDER BY name;
+                """, vars=(style,))
+            for row in cursor.fetchall():
```
I think what @pumazi meant was probably something like:

```
for row in cursor.fetchall():
    result = dict(zip(['name', 'authors', 'revised', 'recipe', 'ident_hash'], row))
    ...
    collections.append({
        'title': row['name'].decode('utf-8'),
        'authors': row['authors'],
        'revised': row['revised'],
        'ident_hash': row['ident_hash'],
        'link': request.route_path('get-content', ident_hash=row['ident_hash']),
        'status': status,
    })
```

The alternative is to use the `psycopg2.extras.DictCursor` (http://initd.org/psycopg/docs/extras.html)

This is still kind of ok because we're not returning too many things in the statement, it would probably make #161 more readable. :D

karenc commented on cnxpublishing/views/admin.py:
```diff
@@ -271,3 +274,86 @@ def admin_edit_site_message_POST(request):
+                WHERE print_style=%s
+                """, vars=(style,))
+            info = cursor.fetchall()
+            if len(info) != 1:
```
I had a look at the [print_style_recipes table](https://github.com/Connexions/cnx-db/blob/master/cnxdb/archive-sql/schema/tables.sql#L285), `print_style` is not unique, `(print_style, tag)` is.  It is possible that this query will return more than one row.  Are we showing the latest one or do we need to show all the versions of it?  If we are just showing one, we probably want to add `ORDER BY revised DESC LIMIT 1` to the SQL statement.

sisiz replied:
I see now. The version of the database I had had print style as unique, but the db has changed since. I think it should show books with all the versions of the print style, so it will get any book with that print style and then also display the recipe and tag in the table.

sisiz replied:
@karenc I think the migration for creating this table did not include this update. https://github.com/Connexions/cnx-db/blob/master/cnxdb/migrations/20170614211640_print-style-recipes.py#L24

karenc replied:
Does it mean If you change `FROM print_style_recipes` to `FROM default_print_style_recipes` it'll work?

karenc commented on cnxpublishing/views/admin.py:
```diff
@@ -280,25 +280,30 @@ def admin_edit_site_message_POST(request):
             cursor.execute("""\
-                SELECT print_style, fileid, recipe_type, revised, tag,
+                SELECT print_style, max(recipe_type),
+                    max(revised), max(tag),
```
I don't think this is what you want.  This query is probably going to select the max `recipe_type`, max `revised` and max `tag` which might be from different rows in the database.

For example consider this table:

```
 print_style  | fileid | recipe_type | tag |            revised            
--------------+--------+-------------+-----+-------------------------------
 ccap-physics |      1 | web         | 1.0 | 2017-08-09 13:14:39.375236-07
 ccap-physics |      2 | pdf         | 2.0 | 2017-08-09 13:14:53.272308-07
```

The query is going to return:

```
 print_style  | max |              max              | max | count 
--------------+-----+-------------------------------+-----+-------
 ccap-physics | web | 2017-08-09 13:14:53.272308-07 | 2.0 |     0
```

I don't know if there's a more efficient way of doing this but this seems to work:

```sql
WITH latest AS (
    SELECT print_style, max(revised) AS revised
        FROM print_style_recipes
        GROUP BY print_style
) SELECT ps.print_style, ps.recipe_type, ps.revised, tag, (
    SELECT count(*)
        FROM latest_modules AS lm
        WHERE lm.print_style = ps.print_style AND lm.portal_type = 'Collection')
    FROM latest JOIN print_style_recipes AS ps ON ps.print_style = latest.print_style AND ps.revised = latest.revised;
```

reedstrm commented:
I'm currently rebasing this against @karenc  cleanup work in content-status

reedstrm commented:
going to close this, and reopen w/ the rebased code later today.
